internal/core/src/common/Array.h

```diff
@@ -0,0 +1,457 @@
+                }
+
+                data_ = new char[size_];
+                for (int i = 0; i < length_; ++i) {
+                    std::copy_n(field_data.string_data().data(i).data(),
+                                field_data.string_data().data(i).size(),
+                                data_ + offsets_[i]);
+                }
+                break;
+            }
+            default: {
+                //                PanicInfo("unsupported element type for array");
+            }
+        }
+    }
+
+    Array(char* data,
+          size_t size,
+          DataType element_type,
+          std::vector<uint64_t>&& element_offsets)
+        : data_(data),
+          size_(size),
+          element_type_(element_type),
+          offsets_(std::move(element_offsets)),
+          length_(element_offsets.size()) {
+    }
+
+    bool
+    operator==(const Array& arr) const {
+        if (element_type_ != arr.element_type_) {
+            return false;
+        }
+        if (length_ != arr.length_) {
+            return false;
+        }
+        for (int i = 0; i < length_; ++i) {
+            switch (element_type_) {
+                case DataType::INT64: {
+                    if (get_data<int64_t>(i) != arr.get_data<int64_t>(i)) {
+                        return false;
+                    }
+                    break;
+                }
+                case DataType::BOOL: {
+                    if (get_data<bool>(i) != arr.get_data<bool>(i)) {
+                        return false;
+                    }
+                    break;
+                }
+                case DataType::DOUBLE: {
+                    if (get_data<double>(i) != arr.get_data<double>(i)) {
+                        return false;
+                    }
+                    break;
+                }
+                case DataType::FLOAT: {
+                    if (get_data<float>(i) != arr.get_data<float>(i)) {
+                        return false;
+                    }
+                    break;
+                }
+                case DataType::INT32:
+                case DataType::INT16:
+                case DataType::INT8: {
+                    if (get_data<int>(i) != arr.get_data<int>(i)) {
+                        return false;
+                    }
+                    break;
+                }
+                case DataType::STRING:
+                case DataType::VARCHAR: {
+                    if (get_data<std::string_view>(i) !=
+                        arr.get_data<std::string_view>(i)) {
+                        return false;
+                    }
+                    break;
+                }
+                default:
+                    PanicInfo("unsupported element type for array");
+            }
+        }
+        return true;
+    }
+
+    template <typename T>
+    T
+    get_data(const int index) const {
+        AssertInfo(
+            index >= 0 && index < length_,
```

Comment:
move this check to outside is better, because check to much is bad for performance

Reply:
There is no way to move it outside, because the number of elements in the array in each row may be different.